### DISCUSSION TITLE
Trojan infected Serial studio

### DISCUSSION THREAD

**Discussion by Vital197** [Serial-Studio/Serial-Studio#109] [Show and tell]
I´m writing this email to rise attention of community because I got trouble in the moment of instalation of the product. In that moment my computer was almost frozen completely and also, I never got to work the program properly. Recently I saw many extra files in every new folders created by Processing. Between these files I saw Serial Studio and I checked this one with https://www.virustotal.com/gui/home/upload this websites. Their antivirus found some trojan Win32. exe https://www.virustotal.com/gui/file/4669b952bcd6ea7056a1603ddaa73d42adfeb2d7e4954bca095ae8b3c5143b3d. I´m sending this email to community to rise their attention.  

Also my computer begun to work very slowly.

**Comment by alex-spataru** [comment]
Hi Vital197!

Thanks a lot for the report. I'm sorry to see that some antivirus programs detected this software as a trojan. I am not sure why this happened. And I'll take the necessary steps to fix this. 

I can assure you that as an open-source developer publishing my hard work with my real name, I don't have any intention of participating in this kind of activities.  

All builds are generated on GitHub's CI server for the following reasons:
- In the case that my system is infected, I don't want to propagate any virus further
- All users can see and verify the build process
- Also, CI builds allow me to ensure that the application is able to compile on *vanilla* setups

You can check the build script for each operating system here:
- [https://github.com/Serial-Studio/Serial-Studio/blob/master/.github/workflows/deploy.yml](https://github.com/Serial-Studio/Serial-Studio/blob/master/.github/workflows/deploy.yml)

The installers are automatically uploaded from GitHub's CI server to the releases page, no download or extra processing is done by me.

I invite you to check the source code, the installer scripts and everything else. Probably this is a false positive emitted by the antivirus because this software interacts with many hardware peripherals (serial, bluetooth, network, etc).

Also, there could be the possibility that the download sources used for generating the installer (e.g. Qt, NSIS, etc) have been compromised, I'll take a look at it later this evening. Please let me know if I can help with anything. 

Finally **please only download Serial Studio from the [GitHub releases page](https://github.com/Serial-Studio/Serial-Studio/releases)**. I don't have any control or say over other software distribution websites!

This includes [SourceForge](https://sourceforge.net/projects/serial-studio.mirror) which for some reason listed my software there and mirrored this repo without my permission or any kind of notification. You must be very careful while downloading open-source software from non-official distribution channels, for example, check this [article](https://arstechnica.com/information-technology/2015/05/sourceforge-grabs-gimp-for-windows-account-wraps-installer-in-bundle-pushing-adware/) about GIMP & SourceForge.

**Reply by alex-spataru** [reply]
Hi! Quick update, I [uploaded a copy](https://www.virustotal.com/gui/file/66ff320370a76ef4c7db740e702ba1d96b07bce6e8956b0c5738c1a9ec4165aa/detection) of `SerialStudio.exe` (not the installer) and no antivirus flagged the software as malicious. The AV flags another file (e.g. a library) or the installer itself. I'll check each file one by one and get in touch with the vendors to see what can be done.

**Reply by alex-spataru** [reply]
Also, @Vital197 if possible, please share a preview of the folders that where affected by Serial Studio. I want to check if that's normal behavior, a bug, or something worse. Again, thanks a lot for your report.

**Reply by alex-spataru** [reply]
As a side note, I think that from now on it should be good practice for me to add the MD5 hashes of each installer/release. If someone wants to lucrate with the work of others, at least we will be able to differentiate between a rouge installer or the ones that are uploaded here.

**Comment by alex-spataru** [comment]
Update, here is the graph of the files associated with the installer that you scanned:

![undefined-4](https://user-images.githubusercontent.com/4225542/170391318-a3361412-9608-4571-9d5f-54356a09bbc1.png)

There are two suspicious things that happen:
- For some reason, the NSIS installer directly contacts an IP from Ireland (`20.82.210.154`) that [seemingly belongs to Microsoft](https://www.abuseipdb.com/check/20.82.210.154). I don't know why this happens.
- The installer creates/bundles a file named [`UN_A.exe`](https://www.virustotal.com/gui/file/50494e097b8749d817b26c2e69f67a502f5cf2182881926b70781a0202e74087) which is flagged by Jiangmin as `Trojan.PSW.Python.fd`.

At first, I though that [`UN_A.EXE`](https://www.virustotal.com/gui/file/50494e097b8749d817b26c2e69f67a502f5cf2182881926b70781a0202e74087) was the uninstaller. However, I uploaded the [`uninstall.exe`](https://www.virustotal.com/gui/file/a5bafa7b1bbb362cda83a6ca0366ef2e66da59aeffb16728dd72550b33cf0243/detection) file from a fresh install and whilst they had similar sizes, they have different hashes and resulted with different scan positives (`Suspicious.low.ml.score` from Trapmine). It is possible that NSIS generates the uninstaller on the fly, however, we need to check if that is the case.

🚩 **Important note:** I also downloaded a copy from GitHub releases and uploaded it to VirusTotal. The hashes of the scanned installer and the one that I just uploaded are exactly the same.

**Reply by alex-spataru** [reply]
*Another update:* I don't want to rescan the `SerialStudio-1.1.7-Windows.exe` file until we ensure that the `makensis.exe` used by GitHub is safe. However, I [uploaded the installer](https://www.virustotal.com/gui/file/361506e868280e218641f201d38cee0a416404659b00b3951989d83ca6d692fd/detection) of an older release (1.1.6) and only Trapmine detects the installer as suspicious. Both Jiangmin and Palo Alto Networks have been updated to avoid detecting false positives.

**Comment by alex-spataru** [comment]
Expanding the details of the [`Un_A.exe`](https://www.virustotal.com/gui/file/50494e097b8749d817b26c2e69f67a502f5cf2182881926b70781a0202e74087) file things get more worrying. It contacts the following IP on the local network:

<img width="351" alt="Screen Shot 2022-05-25 at 19 34 21" src="https://user-images.githubusercontent.com/4225542/170391469-87194337-5e4c-4098-82c0-c28936f546b2.png">

Surprisingly, expanding the installer file with the [Unarchiver](https://theunarchiver.com) does not reveal the `Un_A.exe` file.  The file also contacts another IP address from Microsoft ([`20.62.24.77`](https://www.abuseipdb.com/check/20.62.24.77)). Here is the graph:

<img width="562" alt="Screen Shot 2022-05-25 at 19 45 56" src="https://user-images.githubusercontent.com/4225542/170392495-63453e35-6458-493c-a99f-346ddd5f6adf.png">

And finally, expanding the `192.168.0.53` IP address in the `Un_A.exe` contained in the installer file reveals the following graph:

![undefined-6](https://user-images.githubusercontent.com/4225542/170393101-b9e7bd2a-30f6-46e2-852b-0d8379059d55.png)

@joncloud I am using your GitHub Action to generate the NSIS installer for this project. Do you have any knowledge of this? Or do you know who should we contact in order to clarify this behavior?

**Reply by joncloud** [reply]
I'm not aware of this. The https://github.com/joncloud/makensis-action/ project leverages the existing `makensis` and `makensis.exe` programs on the system (instead of providing them). The project has two simple tests that I've uploaded the results to VirusTotal:

* [test/basic.nsi](https://github.com/joncloud/makensis-action/blob/publish/test/basic.nsi) - [VirusTotal](https://www.virustotal.com/gui/file/9dd5c0527681469277859db92fdd089c9ae0d84e1fc287da259a91f9f12c7733/detection)
* [test/with-plugins.nsi](https://github.com/joncloud/makensis-action/blob/publish/test/with-plugins.nsi) - [VirusTotal](https://www.virustotal.com/gui/file/4a2e651f2974d915296ee0fb4e3b75b0bf86f319473a0a11011fb004a99675e3/detection)

Interestingly enough the VirusTotal results look like they both contact [23.216.147.64](https://www.virustotal.com/gui/ip-address/23.216.147.64), which only Cylance calls out for being unsafe.

I'm unsure about the actual implementations here, however it would seem like it might make sense to get in touch with someone at [NSIS](https://nsis.sourceforge.io/Main_Page), or potentially GitHub directly since they provide the binaries by default on the GitHub Hosted Runners (if you're using them).

**Reply by jp-bennett** [reply]
https://otx.alienvault.com/indicator/hostname/shell.windows.com That IP seems to be part of a Microsoft service to lookup how to open a given file type. As per https://devblogs.microsoft.com/oldnewthing/20090113-00/?p=19513

**Reply by alex-spataru** [reply]
@joncloud Thanks a lot for your response! I'll also mention @kichik (I believe that this is the account of the primary NSIS developer) to see if he can provide us with a little insight of this issue. 

@kichik I already read the [NSIS false positives page](https://nsis.sourceforge.io/NSIS_False_Positives), and will take action ASAP. However, I would like to know if this should be normal behavior. The implications of having a rouge `makensis.exe` installed on GitHub's servers kind of scares me, since it would not only affect this project...

**Reply by kichik** [reply]
1. The strongest indication this is a false positive are the generic "trojan" names. Also the obscure AVs that "caught" it. 
2. Looks like GitHub installs makensis as non of your workflows do it explicitly. So it's probably safe. 
3. You can check the version with `makensis /HDRINFO` to make it easier to compare hashes with the right exe. 
4. Those Microsoft IPs are probably just Windows telemetry. I'd bet you see those everywhere but I haven't checked. Either way confirming makensis.exe hash should resolve it.

**Reply by alex-spataru** [reply]
@kichik Thanks a lot for your insights. I just modified the GitHub action to execute `makensis /HDRINFO` and the output is exactly the same as the NSIS installer downloaded from the official website:

**makensis (local install in a virtual machine)**

![nsis](https://user-images.githubusercontent.com/4225542/170763453-00efb47d-52eb-4b70-a505-41b49cd0bbf0.PNG)

**makensis (GitHub server)**

<img width="1128" alt="Screen Shot 2022-05-27 at 12 42 11" src="https://user-images.githubusercontent.com/4225542/170763513-5baec1f7-43b1-4fba-91ee-8e09df636aee.png">

I'll get in touch with the AV developers to report a false positive. @Vital197 @joncloud thanks a lot for your messages & insights!

**Reply by kichik** [reply]
After thinking about it some more, you probably want to confirm the hashes of all the files under Stubs and Plugins too if you have concerns. The uninstaller code comes from Stubs and packaged up by makensis. Whatever nsh files you're using probably could use verification too. I doubt you will find anything because this looks exactly like all the other false positives. But if you want to be completely sure, it can't hurt. Maybe with something like https://idrassi.github.io/DirHash/

**Reply by Vital197** [reply]
Hi Alex and community,

thank you so much for all your replies and help.  Sorry for my delay I got so busy and also sad with this virus which I got from the installer. Again I want to point out why I thought that problem was related to Serial Studio. First of all, my computer was frozen in time of instalation Serial Studio, I can tell that it was something unusual. Secondly, in new created folders for new Processing sketch I have file named "Serial Studio 1.1.7.exe".  That are two reasones why I thought that problem in the installer. 
Almost one week I try two fight with this problem and it still there. In this moment I can not continue with my Processing and Arduino project.
Usualy I try to download files from safe sources and this case I believe it was dowloaded from gitHub But believeme that I really careful with that and I have enough experience with it. Also I can suspect my processing installer but that I got from processing.org.
I tried to check my sistem with Kaspersky antivirus but it can not find anything. I will try to check with external disk and with different antiviruses. 
I try to put here screeenshot of infected Processing folder. In this folder you can see bunch of diferrent files which does not have relation to *.pde file.  
![image](https://user-images.githubusercontent.com/106120559/172719060-74c2c412-302d-4e5f-b21c-08ea53ea3c79.png)
![image](https://user-images.githubusercontent.com/106120559/172719177-e3ab83e7-b874-47dd-9bc5-b76f36589ee4.png)

 
![Untitled](https://user-images.githubusercontent.com/106120559/172719678-99794efb-5639-4cd9-9446-32605bd42fe2.png)
 
I´m not sure that I was able to dowload  my screenshort of infected folders. Please let me know if that case and we will try to find some another way to show you infected folders.

**Reply by Vital197** [reply]
I see that it works. So now you can see my infected folders.

**Reply by alex-spataru** [reply]
Hi @Vital197! Again sorry for the bad experience, I have checked the installers, talked with AV vendors and we have confirmed that indeed the NSIS installer was detected as a false positive. [Here](https://nsis.sourceforge.io/NSIS_False_Positives) is a more detailed explanation for this from the official website of NSIS. 

>  First of all, my computer was frozen in time of instalation Serial Studio, I can tell that it was something unusual.

I think that this could be related to the Microsoft Visual C++ redistributable installer that is [called automatically](https://github.com/Serial-Studio/Serial-Studio/blob/master/deploy/windows/nsis/setup.nsi#L115) during the installation of Serial Studio. The Visual C++ redistributable installs the compiler runtime needed to be able to run the MSVC-compiled version of Serial Studio. 

I don't know if using Windows 7 would affect the installation of the Visual C++ redistributable package, as [this operating system is not supported anymore by Microsoft](https://support.microsoft.com/en-us/windows/windows-7-support-ended-on-january-14-2020-b75d4580-2cc7-895a-2c9c-1466d9a53962). I think that there could be a chance that the Visual C++ redistributable installer got stuck during the installation process.

I have already done tests on a Windows 7 SP1 (build 7601) VM and I did not experience any issues. I think that there is a probability that your computer catched the virus from elsewhere and somehow it placed the Serial Studio installer (plus a bunch of other files) in your Processing folder.

<img width="1407" alt="Screen Shot 2022-06-08 at 22 55 34" src="https://user-images.githubusercontent.com/4225542/172760840-b2999a4b-4f87-4d6e-bccb-679f2937c7ff.png">

I would really recommend you to upgrade to Windows 10 or Windows 11, since Windows 7 (unfortunately) does not receive security updates anymore, here is Microsoft's stance:

> If you continue to use Windows 7 now that support has ended, your PC will still work, but it will be more vulnerable to security risks and viruses. Your PC will continue to start and run, but will no longer receive software updates, including security updates, from Microsoft.

I use Windows 7 for interacting with several proprietary motor controllers/drivers for my thesis. However, I do this from a VM with read-only access to my primary operating system. This way, I can still run legacy software, and interact with specialized hardware while mitigating any possible security issues.

**Reply by alex-spataru** [reply]
Also, if needed. Feel free to send me a private e-mail to schedule a Zoom call to check what is happening with your Serial Studio installation. Spanish is my native language :)
